builder/dispatchers.go

```diff
@@ -556,3 +613,43 @@ func volume(b *builder, args []string, attributes map[string]bool, original stri
 	}
 	return nil
 }
+
+// ARG name=[value]
+//
+// Adds the variable foo to the trusted list of variables that can be passed
+// to builder using the --build-arg flag for expansion/subsitution or passing to 'run'.
+// Dockerfile author may optionally set a default value of this variable.
+func arg(b *builder, args []string, attributes map[string]bool, original string) error {
+	if len(args) != 1 {
+		return fmt.Errorf("ARG requires exactly one argument")
+	}
+
+	var (
+		name           string
+		value          string
+		isValuePresent bool
+	)
+
+	arg := args[0]
+	// 'arg' can just be a name or name-value pair
+	if strings.Contains(arg, "=") {
+		parts := strings.SplitN(arg, "=", 2)
+		name = parts[0]
+		value = parts[1]
+		isValuePresent = true
+	} else {
+		name = arg
+		isValuePresent = false
+	}
+	// add the arg to allowed list of build-time args from this step on.
+	b.allowedBuildArgs[name] = true
+
+	// If there is a default value associated with this arg then add it to the
+	// b.buildArgs if one is not already passed to the builder. The args passed
+	// to builder override the defaut value of 'arg'.
+	if _, ok := b.buildArgs[name]; !ok && isValuePresent {
```

Comment:
Yes, it doesn't matter as ultimately it results in the same behavior wrt a bash script like check which works the same for empty string or unset variables.

Here I was trying to distinguish the two scenarios, let me know if it makes sense. I think it might come handy when user wants to make sure we don't introduce variables in the `RUN` environment until they are explicitly passed using `--build-arg` flag:

```
ARG key=""  // in this case I treat it as **allowed with a default**
ARG key1=   // in this case I treat it as **allowed but not set** until it is explicitly passed as build-arg

RUN env     >> this will print 'key' but not 'key1' in the environment
```

But if you see this complicates things, I can change it to always set value to empty string for all *defined but not set* `ARG`s.